vortex-layout/src/layouts/chunked/reader.rs

```diff
@@ -101,35 +103,32 @@ impl ChunkedReader {
     }
 
     pub(crate) async fn pruning_mask(&self, expr: &ExprRef) -> VortexResult<Option<BooleanBuffer>> {
-        if let Some(mask) = self
+        let cell = self
             .pruning_result
-            .read()
+            .write()
             .map_err(|_| vortex_err!("poisoned lock"))?
-            .get(expr)
-        {
-            return Ok(mask.clone());
-        }
-        let pruning_predicate = PruningPredicate::try_new(expr);
-        let res = if let Some(stats_table) = self.stats_table().await? {
-            if let Some(predicate) = pruning_predicate {
-                predicate
-                    .evaluate(stats_table.array())?
-                    .map(|mask| mask.into_bool())
-                    .transpose()?
-                    .map(|mask| mask.boolean_buffer())
+            .entry(expr.clone())
+            .or_insert(Arc::new(OnceCell::new()))
```

Comment:
Is there a variant of this that takes a closure so that we don't greedily create a new one

Reply:
we're holding the write lock here, so it shouldn't happen.